qiskit/transpiler/passes/layout/disjoint_utils.py

```diff
@@ -0,0 +1,152 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""This module contains common utils for disjoint coupling maps."""
+
+from collections import defaultdict
+from typing import List, Callable, TypeVar, Dict
+import uuid
+
+import rustworkx as rx
+
+from qiskit.circuit import Qubit, Barrier, Clbit
+from qiskit.dagcircuit.dagcircuit import DAGCircuit
+from qiskit.dagcircuit.dagnode import DAGOutNode
+from qiskit.transpiler.coupling import CouplingMap
+from qiskit.transpiler.exceptions import TranspilerError
+from qiskit.transpiler.passes.layout import vf2_utils
+
+T = TypeVar("T")
+
+
+def run_pass_over_connected_components(
+    dag: DAGCircuit,
+    coupling_map: CouplingMap,
+    run_func: Callable[[DAGCircuit, CouplingMap], T],
+) -> List[T]:
+    """Run a transpiler pass inner function over mapped components."""
+    cmap_components = coupling_map.connected_components()
+    # If graph is connected we only need to run the pass once
+    if len(cmap_components) == 1:
+        return [run_func(dag, cmap_components[0])]
+    dag_components = separate_dag(dag)
+    mapped_components = map_components(dag_components, cmap_components)
+    out_component_pairs = []
+    for cmap_index, dags in mapped_components.items():
+        out_dag = dag_components[dags.pop()]
```

Comment:
Might be helpful to have a comment here that explains we're just picking the first (and possibly only) DAG in the bucket and then composing all other circuits in the bucket with this one. I was a bit confused at first since I didn't realize we would combine some of the split DAGs before running the pass (though it's obvious in post).

Reply:
Updated in: https://github.com/Qiskit/qiskit-terra/pull/9802/commits/5df53a7003412e9d4c5b60ae6816b02a564f6c07